<a href="https://colab.research.google.com/github/richieyuyongpoh/AI4Fun/blob/main/Use_Case_Customer_Churn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Use Case - Customer Churn**



Prepared by:

Dr. Yong Poh Yu


LinkedIn: [Yong Poh Yu
](https://www.linkedin.com/in/yong-poh-yu/)

Github: [Practical Python Series](https://github.com/richieyuyongpoh/Practical_Python_series)

## Hands-On Practice



Please upload the [churn.txt](https://github.com/richieyuyongpoh/Practical_Python_series/blob/main/churn.txt) into current session. Take note that this dataset was downloaded from [Kaggle](https://www.kaggle.com/c/customer-churn-prediction/data). The metadata can be found [here](https://www.kaggle.com/c/customer-churn-prediction/discussion/34397) 

### Initialization

1: Read the csv and save as a dataframe/table

In [1]:

import pandas as pd

data = pd.read_table('https://raw.githubusercontent.com/richieyuyongpoh/Practical_Python_series/main/churn.txt', index_col=False,  sep=',',skipinitialspace=True)

data

st  acclen  arcode     phnum intplan  ... timin  tical  tichar  ncsc   label
0     KS     128     415  382-4657      no  ...  10.0      3    2.70     1  False.
1     OH     107     415  371-7191      no  ...  13.7      3    3.70     1  False.
2     NJ     137     415  358-1921      no  ...  12.2      5    3.29     0  False.
3     OH      84     408  375-9999     yes  ...   6.6      7    1.78     2  False.
4     OK      75     415  330-6626     yes  ...  10.1      3    2.73     3  False.
...   ..     ...     ...       ...     ...  ...   ...    ...     ...   ...     ...
3328  AZ     192     415  414-4276      no  ...   9.9      6    2.67     2  False.
3329  WV      68     415  370-3271      no  ...   9.6      4    2.59     3  False.
3330  RI      28     510  328-8230      no  ...  14.1      6    3.81     2  False.
3331  CT     184     510  364-6381     yes  ...   5.0     10    1.35     2  False.
3332  TN      74     415  400-4344      no  ...  13.7      4    3.70     0  False.

[3333 rows x 21 columns]

2: As the `phnum` is the `primary key` or `id` for this table/dataset, remove it. 

In [2]:
data = data.drop('phnum',axis=1)

data.head()

st  acclen  arcode intplan voice  ...  timin  tical  tichar  ncsc   label
0  KS     128     415      no   yes  ...   10.0      3    2.70     1  False.
1  OH     107     415      no   yes  ...   13.7      3    3.70     1  False.
2  NJ     137     415      no    no  ...   12.2      5    3.29     0  False.
3  OH      84     408     yes    no  ...    6.6      7    1.78     2  False.
4  OK      75     415     yes    no  ...   10.1      3    2.73     3  False.

[5 rows x 20 columns]

3: Get the simple descriptive statistics

In [3]:

data.info()
data.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   st          3333 non-null   object 
 1   acclen      3333 non-null   int64  
 2   arcode      3333 non-null   int64  
 3   intplan     3333 non-null   object 
 4   voice       3333 non-null   object 
 5   nummailmes  3333 non-null   int64  
 6   tdmin       3333 non-null   float64
 7   tdcal       3333 non-null   int64  
 8   tdchar      3333 non-null   float64
 9   temin       3333 non-null   float64
 10  tecal       3333 non-null   int64  
 11  tecahr      3333 non-null   float64
 12  tnmin       3333 non-null   float64
 13  tn cal      3333 non-null   int64  
 14  tnchar      3333 non-null   float64
 15  timin       3333 non-null   float64
 16  tical       3333 non-null   int64  
 17  tichar      3333 non-null   float64
 18  ncsc        3333 non-null   int64  
 19  label       3333 non-null  

acclen       arcode  ...       tichar         ncsc
count  3333.000000  3333.000000  ...  3333.000000  3333.000000
mean    101.064806   437.182418  ...     2.764581     1.562856
std      39.822106    42.371290  ...     0.753773     1.315491
min       1.000000   408.000000  ...     0.000000     0.000000
25%      74.000000   408.000000  ...     2.300000     1.000000
50%     101.000000   415.000000  ...     2.780000     1.000000
75%     127.000000   510.000000  ...     3.270000     2.000000
max     243.000000   510.000000  ...     5.400000     9.000000

[8 rows x 16 columns]

4: Define the Features , `X`

In [4]:
X = data.drop('label',axis=1)


X.head()

st  acclen  arcode intplan voice  ...  tnchar  timin  tical  tichar  ncsc
0  KS     128     415      no   yes  ...   11.01   10.0      3    2.70     1
1  OH     107     415      no   yes  ...   11.45   13.7      3    3.70     1
2  NJ     137     415      no    no  ...    7.32   12.2      5    3.29     0
3  OH      84     408     yes    no  ...    8.86    6.6      7    1.78     2
4  OK      75     415     yes    no  ...    8.41   10.1      3    2.73     3

[5 rows x 19 columns]

5: Remove the categorical variables that are written in texts, such as `st`, `intplan`, `voice`.
<br><br>
Note: in the later part, we will pre-process these categorical variables and then include them for the modelling. Refer to step [#14](https://colab.research.google.com/drive/1ndH11fsmbKMX7S-xtRMqAOCBcXnwxLOs?authuser=1#scrollTo=2zdN6QPfdzON&line=3&uniqifier=1)

In [5]:
X = X.drop(['st','intplan','voice'],axis=1)


X.head()

acclen  arcode  nummailmes  tdmin  tdcal  ...  tnchar  timin  tical  tichar  ncsc
0     128     415          25  265.1    110  ...   11.01   10.0      3    2.70     1
1     107     415          26  161.6    123  ...   11.45   13.7      3    3.70     1
2     137     415           0  243.4    114  ...    7.32   12.2      5    3.29     0
3      84     408           0  299.4     71  ...    8.86    6.6      7    1.78     2
4      75     415           0  166.7    113  ...    8.41   10.1      3    2.73     3

[5 rows x 16 columns]

6: define the label, `y`

In [6]:
y = data['label']
y.head()

0    False.
1    False.
2    False.
3    False.
4    False.
Name: label, dtype: object

7: For the quick start, let us ignore pre-processing steps and directly move on to the machine learning (training & testing) steps.

Let us start with `train-test-split()`

In [7]:
from sklearn.model_selection import train_test_split
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y)

In [8]:
Xtrain.head()

acclen  arcode  nummailmes  tdmin  ...  timin  tical  tichar  ncsc
2616     165     415          33  111.6  ...   16.0      3    4.32     0
353      117     415           0  181.5  ...   14.7      9    3.97     2
3154      73     415          47  173.7  ...    6.3      3    1.70     2
31        75     510           0  226.1  ...   10.3      5    2.78     1
2820      74     510          27  154.1  ...   13.2      2    3.56     4

[5 rows x 16 columns]

In [9]:
ytrain.head()

2616    False.
353     False.
3154    False.
31      False.
2820    False.
Name: label, dtype: object

### *k*-nearest neigbours (*k*-NN model)

8: Apply *k*-NN model. 

In [10]:

from sklearn.neighbors import KNeighborsClassifier


knn = KNeighborsClassifier()
knn.fit(Xtrain, ytrain)
ypred = knn.predict(Xtest)




9: Observe the result

In [11]:
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(ytest, ypred))
print()
print()
print(classification_report(ytest, ypred))

[[697  15]
 [ 89  33]]


              precision    recall  f1-score   support

      False.       0.89      0.98      0.93       712
       True.       0.69      0.27      0.39       122

    accuracy                           0.88       834
   macro avg       0.79      0.62      0.66       834
weighted avg       0.86      0.88      0.85       834



### Support Vector Machine (SVM)

10: Apply SVM and observe the results

In [12]:
from sklearn.svm import SVC

svc = SVC()
svc.fit(Xtrain, ytrain)
ypred = svc.predict(Xtest)

print(confusion_matrix(ytest, ypred))
print()
print()
print(classification_report(ytest, ypred))

[[712   0]
 [122   0]]


              precision    recall  f1-score   support

      False.       0.85      1.00      0.92       712
       True.       0.00      0.00      0.00       122

    accuracy                           0.85       834
   macro avg       0.43      0.50      0.46       834
weighted avg       0.73      0.85      0.79       834



/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### Logistic Regression

11: Apply Logistic Regression and observe the results

In [13]:
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression()

logreg.fit(Xtrain, ytrain)
ypred = logreg.predict(Xtest)

print(confusion_matrix(ytest, ypred))
print()
print()
print(classification_report(ytest, ypred))

[[707   5]
 [115   7]]


              precision    recall  f1-score   support

      False.       0.86      0.99      0.92       712
       True.       0.58      0.06      0.10       122

    accuracy                           0.86       834
   macro avg       0.72      0.53      0.51       834
weighted avg       0.82      0.86      0.80       834



/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


### Gaussian Naive Bayes

12: Apply NB and observe the results

In [14]:
from sklearn.naive_bayes import GaussianNB

nb = GaussianNB()
nb.fit(Xtrain, ytrain)
ypred = nb.predict(Xtest)

print(confusion_matrix(ytest, ypred))
print()
print()
print(classification_report(ytest, ypred))


[[682  30]
 [ 72  50]]


              precision    recall  f1-score   support

      False.       0.90      0.96      0.93       712
       True.       0.62      0.41      0.50       122

    accuracy                           0.88       834
   macro avg       0.76      0.68      0.71       834
weighted avg       0.86      0.88      0.87       834



### Random Forest

13: Apply RF and observe the results

In [15]:
from sklearn.ensemble import RandomForestClassifier

RandomForest = RandomForestClassifier()
RandomForest.fit(Xtrain, ytrain)
ypred = RandomForest.predict(Xtest)

print(confusion_matrix(ytest, ypred))
print()
print()
print(classification_report(ytest, ypred))

[[700  12]
 [ 55  67]]


              precision    recall  f1-score   support

      False.       0.93      0.98      0.95       712
       True.       0.85      0.55      0.67       122

    accuracy                           0.92       834
   macro avg       0.89      0.77      0.81       834
weighted avg       0.92      0.92      0.91       834



It seems Random Forest gives the best results.  

### Visualization Dashboard

Please click the [link](https://www.richieyyptutorialpage.com/demo-python-series/machine-learning-with-scikit-learn) provided by the trainer and follow the instruction. 

### Improvements

14: Apply label encoder to the categorical variables

Refer to step [#5](https://colab.research.google.com/drive/1ndH11fsmbKMX7S-xtRMqAOCBcXnwxLOs?authuser=1#scrollTo=JLHTAof1Ps6f&line=3&uniqifier=1)

In [18]:
from sklearn.preprocessing import LabelEncoder

data = pd.read_table('https://raw.githubusercontent.com/richieyuyongpoh/Practical_Python_series/main/churn.txt', index_col=False,  sep=',',skipinitialspace=True)
data = data.drop('phnum',axis=1)

labelencoder = LabelEncoder()
data['st'] = labelencoder.fit_transform(data['st'])
data['intplan'] = labelencoder.fit_transform(data['intplan'])
data['voice'] = labelencoder.fit_transform(data['voice'])

data.head()

st  acclen  arcode  intplan  voice  ...  timin  tical  tichar  ncsc   label
0  16     128     415        0      1  ...   10.0      3    2.70     1  False.
1  35     107     415        0      1  ...   13.7      3    3.70     1  False.
2  31     137     415        0      0  ...   12.2      5    3.29     0  False.
3  35      84     408        1      0  ...    6.6      7    1.78     2  False.
4  36      75     415        1      0  ...   10.1      3    2.73     3  False.

[5 rows x 20 columns]

15: Repeat #4, #6, and #7

In [19]:
X = data.drop('label',axis=1)

y = data['label']

Xtrain, Xtest, ytrain, ytest = train_test_split(X, y)

16: Repeat #13 

In [20]:
from sklearn.ensemble import RandomForestClassifier

RandomForest = RandomForestClassifier()
RandomForest.fit(Xtrain, ytrain)
ypred = RandomForest.predict(Xtest)

print(confusion_matrix(ytest, ypred))
print()
print()
print(classification_report(ytest, ypred))

[[709   8]
 [ 37  80]]


              precision    recall  f1-score   support

      False.       0.95      0.99      0.97       717
       True.       0.91      0.68      0.78       117

    accuracy                           0.95       834
   macro avg       0.93      0.84      0.87       834
weighted avg       0.94      0.95      0.94       834



### What are the leading factors that affect churn rate?

17: check the feature importance

In [21]:
RandomForest.feature_importances_

array([0.02226565, 0.03006284, 0.00819884, 0.07939012, 0.02568825,
       0.02488794, 0.13927275, 0.02826977, 0.13621662, 0.06029316,
       0.0290536 , 0.06101671, 0.03602491, 0.02925906, 0.03514876,
       0.04221867, 0.04460004, 0.04211422, 0.1260181 ])

In [22]:
list(X.columns)

['st',
 'acclen',
 'arcode',
 'intplan',
 'voice',
 'nummailmes',
 'tdmin',
 'tdcal',
 'tdchar',
 'temin',
 'tecal',
 'tecahr',
 'tnmin',
 'tn cal',
 'tnchar',
 'timin',
 'tical',
 'tichar',
 'ncsc']

In [23]:
important_factors = pd.DataFrame({'Factor': list(X.columns), 'Importance': RandomForest.feature_importances_})
important_factors

Factor  Importance
0           st    0.022266
1       acclen    0.030063
2       arcode    0.008199
3      intplan    0.079390
4        voice    0.025688
5   nummailmes    0.024888
6        tdmin    0.139273
7        tdcal    0.028270
8       tdchar    0.136217
9        temin    0.060293
10       tecal    0.029054
11      tecahr    0.061017
12       tnmin    0.036025
13      tn cal    0.029259
14      tnchar    0.035149
15       timin    0.042219
16       tical    0.044600
17      tichar    0.042114
18        ncsc    0.126018

In [24]:
important_factors.sort_values(by=['Importance'], ascending=False,inplace=True)
important_factors

Factor  Importance
6        tdmin    0.139273
8       tdchar    0.136217
18        ncsc    0.126018
3      intplan    0.079390
11      tecahr    0.061017
9        temin    0.060293
16       tical    0.044600
15       timin    0.042219
17      tichar    0.042114
12       tnmin    0.036025
14      tnchar    0.035149
1       acclen    0.030063
13      tn cal    0.029259
10       tecal    0.029054
7        tdcal    0.028270
4        voice    0.025688
5   nummailmes    0.024888
0           st    0.022266
2       arcode    0.008199

### Conclusion



Random Forest gives the best result. By doing some enhancement steps, the accuracy increases to 97%++ while recall (for class 1 - churn ) increases to 80%++ . 